rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,519 @@
+
+## Migrate existing Nixpkgs comments
+
+Reformatting existing doc-comments in Nixpkgs.
+
+Action points:
+
+- [ ] Change comments to markdown.
+- [ ] Migrate nixdoc 'argument documentation' format.
+
+### Migration Example
+
+The following shows one of the many possible ways to migrate the current Nixpkgs `lib` comments.
+
+> We managed to partially automate this effort with a [codemod](https://github.com/nix-community/docnix/tree/3c0531cb5b4c9f3e9069b73d19e6c4be8508d905/codemod)
+
+> Note: The current `nixdoc` feature 'Function arguments' uses a different mental model than this RFC. Arguments must now be explicitly documented inside of the lambda documentation.
+
+`lib/attrsets.nix (old format)`
+````nix
+/* Filter an attribute set by removing all attributes for which the
+   given predicate return false.
+
+   Example:
+     filterAttrs (n: v: n == "foo") { foo = 1; bar = 2; }
+     => { foo = 1; }
+
+   Type:
+     filterAttrs :: (String -> Any -> Bool) -> AttrSet -> AttrSet
+*/
+filterAttrs =
+  # Predicate taking an attribute name and an attribute value, which returns `true` to include the attribute or `false` to exclude the attribute.
+  pred:
+  # The attribute set to filter
+  set:
+  listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
+````
+
+->
+
+`lib/attrsets.nix (new format)`
+````nix
+/**
+  Filter an attribute set by removing all attributes for which the
+  given predicate return false.
+
+  # Example
+
+  ```nix
+  filterAttrs (n: v: n == "foo") { foo = 1; bar = 2; }
+  => { foo = 1; }
+  ```
+
+  # Type
+
+  ```
+  filterAttrs :: (String -> Any -> Bool) -> AttrSet -> AttrSet
+  ```
+
+  # Arguments
+
+  - [pred] Predicate taking an attribute name and an attribute value, which returns `true` to include the attribute, or `false` to exclude the attribute.
+  - [set] The attribute set to filter
+*/
+filterAttrs =
+  pred:
+  set:
+  listToAttrs (concatMap (name: let v = set.${name}; in if pred name v then [(nameValuePair name v)] else []) (attrNames set));
+````
+
+## Tooling
+
+All current rendering tooling solutions should support displaying the specified doc-comment format.
+
+Currently, at least:
+
+### Nixpkgs Manual renderer
+
+The current Nixpkgs manual needs to be adapted to this change.
+
+We expect changes in the following to be necessary:
+
+- [nixos_render_docs](https://github.com/NixOS/Nixpkgs/tree/e4082efedb483eb0478c3f014fa851449bca43f9/pkgs/tools/nix/nixos-render-docs/src) 
+- [nixdoc](https://github.com/nix-community/nixdoc)
+
+## Native support in Nix
+
+- `NixOS/nix` should implement native support for doc-comments.
+
+Primarily because the *Nix* language requires access to the actual evaluator for building correct relations between doc-comment and the node they document.
+
+> Note: We considered implementation details, but specifying those is out of the scope of this RFC.
```

Comment:
> This is risky to a degree.

You cannot plan everything right? I think it is okay if there are limitations to the implementation. Although they could be resolved over time.

Reply:
> You cannot plan everything right? I think it is okay if there are limitations to the implementation. Although they could be resolved over time.

@hsjobeki you keep saying "the implementation" when talking about the Nix language's grammar and semantics.  This is not what most people mean when they say "the implementation".

Reply:
> you keep saying "the implementation" when talking about the Nix language's grammar and semantics.

I think we misunderstand each other.

By implementation i mean any tool or feature that implements against this RFC. 

Afaik we don't change any grammar or semantics of the language. 

There might be changes in one of the implementations of the language. Being the lexer of c++ nix because this implementation of the language doesn't keep comments in the ast. But that is an internal design question of c++ nix.
Tvix doesn't have this problem because they use rnix, which keeps trivial tokens in the ast.

Big problem is because nix doesn't have a language specification any change in the c++ lexer or parser directly affects a potential language specification because the current source of the language is only our holy reference implementation.

I am trying to find a compromise between:

- removing the whole "nix should implement this" point
- nix should implement against this rfc in the future, but we cannot specify how and when.

Do you have any suggestions, that resolve our problem?

Reply:
Just removed the statement. I'll mark this as resolved.